File: architecture-skills/design-patterns/examples/bridge/bridge.java
// Bridge pattern in Java.
//
// Shapes (the abstraction) hold a Renderer (the implementation) so both
// hierarchies grow independently — the same split JDBC uses between the
// API you call and the driver that implements it.
//
// The public class is lowercase so the filename matches the pattern name.

public class bridge {

    interface Renderer {
        String circle(double radius);
        String rectangle(double width, double height);
    }

    static class VectorRenderer implements Renderer {
        public String circle(double radius) {
            return String.format("<circle r=%.1f/>", radius);
        }
        public String rectangle(double width, double height) {
            return String.format("<rect w=%.1f h=%.1f/>", width, height);
        }
    }

    static class RasterRenderer implements Renderer {
        public String circle(double radius) {
            return String.format("pixels for a circle of radius %.1f", radius);
        }
        public String rectangle(double width, double height) {
            return String.format("pixels for a %.1fx%.1f rectangle", width, height);
        }
    }

    abstract static class Shape {
        protected final Renderer renderer;

        Shape(Renderer renderer) {
            this.renderer = renderer;
        }

        abstract String draw();
    }

    static class Circle extends Shape {
        private final double radius;

        Circle(Renderer renderer, double radius) {
            super(renderer);
            this.radius = radius;
        }

        String draw() {
            return renderer.circle(radius);
        }
    }

    static class Rectangle extends Shape {
        private final double width;
        private final double height;

        Rectangle(Renderer renderer, double width, double height) {
            super(renderer);
            this.width = width;
            this.height = height;
        }

        String draw() {
            return renderer.rectangle(width, height);
        }
    }

    public static void main(String[] args) {
        Renderer[] renderers = {new VectorRenderer(), new RasterRenderer()};
        for (Renderer renderer : renderers) {
            Shape[] shapes = {new Circle(renderer, 5.0), new Rectangle(renderer, 3.0, 4.0)};
            for (Shape shape : shapes) {
                System.out.println(shape.draw());
            }
        }
    }
}

File: architecture-skills/design-patterns/examples/bridge/bridge.cs
// Bridge pattern in C#.
//
// Shapes (the abstraction) hold an IRenderer (the implementation) so
// both hierarchies grow independently — the same split ADO.NET uses
// between DbConnection and each provider's implementation.

using System;

interface IRenderer
{
    string Circle(double radius);
    string Rectangle(double width, double height);
}

class VectorRenderer : IRenderer
{
    public string Circle(double radius) => $"<circle r={radius:F1}/>";
    public string Rectangle(double width, double height) => $"<rect w={width:F1} h={height:F1}/>";
}

class RasterRenderer : IRenderer
{
    public string Circle(double radius) => $"pixels for a circle of radius {radius:F1}";
    public string Rectangle(double width, double height) => $"pixels for a {width:F1}x{height:F1} rectangle";
}

abstract class Shape
{
    protected readonly IRenderer Renderer;

    protected Shape(IRenderer renderer) => Renderer = renderer;

    public abstract string Draw();
}

class Circle : Shape
{
    private readonly double _radius;

    public Circle(IRenderer renderer, double radius) : base(renderer) => _radius = radius;

    public override string Draw() => Renderer.Circle(_radius);
}

class Rectangle : Shape
{
    private readonly double _width;
    private readonly double _height;

    public Rectangle(IRenderer renderer, double width, double height) : base(renderer)
    {
        _width = width;
        _height = height;
    }

    public override string Draw() => Renderer.Rectangle(_width, _height);
}

class Program
{
    static void Main()
    {
        var renderers = new IRenderer[] { new VectorRenderer(), new RasterRenderer() };
        foreach (var renderer in renderers)
        {
            var shapes = new Shape[] { new Circle(renderer, 5.0), new Rectangle(renderer, 3.0, 4.0) };
            foreach (var shape in shapes)
            {
                Console.WriteLine(shape.Draw());
            }
        }
    }
}
